core/src/main/java/jenkins/fingerprints/FingerprintStorage.java

```diff
@@ -80,4 +86,39 @@ public static FingerprintStorage get() {
      */
     public abstract boolean isReady();
 
+    /**
+     * Iterates a set of fingerprints, and cleans them up.
+     * This method is called periodically by {@link hudson.model.FingerprintCleanupThread}.
+     * For reference, see {@link FileFingerprintStorage#iterateAndCleanupFingerprints(TaskListener)}
+     * For cleaning up the fingerprint {@link #cleanFingerprint(Fingerprint, TaskListener)} may be used.
+     */
+    public abstract void iterateAndCleanupFingerprints(TaskListener taskListener);
+
+    /**
+     * This method performs the cleanup of the given fingerprint.
+     */
+    public boolean cleanFingerprint(@NonNull Fingerprint fingerprint, TaskListener taskListener) {
+        try {
+            if (!fingerprint.isAlive() && fingerprint.getFacetBlockingDeletion() == null) {
+                taskListener.getLogger().println("deleting obsolete " + fingerprint.toString());
+                Fingerprint.delete(fingerprint.getHashString());
+                return true;
+            } else {
+                if (!fingerprint.isAlive()) {
+                    FingerprintFacet deletionBlockerFacet = fingerprint.getFacetBlockingDeletion();
+                    taskListener.getLogger().println(deletionBlockerFacet.getClass().getName() + " created on " +
```

Comment:
Hmmm...this seems duplicate of https://github.com/jenkinsci/jenkins/pull/4817/files#r445623491

Reply:
Also below as well

Reply:
It's called in different places,
FileFingerprint has it's own implementation, the redis plugin is calling this method

Reply:
Yes, that is correct. This is not duplicate, it has slight changes. The one in `FingerprintStorage` lays out a basic implementation for plugin developers. File system storage overrides it and uses a slightly different implementation instead which can delete files directly.